enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1016 @@
+addresses from a configured range, so I don't need to manage my own IP pools,
+DHCP server, or static IP configuration.
+- As an Openshift Virtualization user, I want to migrate (one of) my VMs to
+another node, preserving the networking configuration, so the IP-aware
+applications running in the VM are not impacted.
+- As an Openshift Virtualization user, I want to start Virtual Machines
+(currently stopped) with the original VM network configuration, so the IP-aware
+applications running in the VM are not impacted.
+- As an Openshift network administrator, I want to decide which networks can be
+subject to having persistent IP allocations surviving VM migration, and VM stop.
+
+### Goals
+
+- Manage Openshift Virtual Machines IP addresses from within the SDN solution
+provided by OVN-Kubernetes.
+- IPv4 range can be defined on OVN Kubernetes secondary network
+- VMs are able to get a IP from this subnet
+- The IP sticks to a VM during live-migration
+- The IP sticks to a VM even if it is temporarily shut down
+
+### Non-Goals
+
+- IPv6 support is not required.
+- This enhancement's scope is limited to secondary networks - i.e. the default
+  cluster network's IP allocation should still follow the pod lifecycle.
+
+## Proposal
+
+### Workflow Description
+
+#### Creating a Virtual Machine
+The user initially requests a virtual machine (VM) to be created. If they
+desire their VM to have "sticky IPs" - i.e. the IP allocations to outlive the
+pod the VM runs on - the following requirements must be met:
+
+- the KubeVirt CR must have the required feature gate enabled: e.g.
+  `VMPersistentIPs`
+- the NAD holding the configuration for the secondary network must allow for
+persistent IPs - `{..., "allowPersistentIPs": true, ...}`
+
+If the KubeVirt feature gate is enabled, KubeVirt will create the `IPAMClaims`
+for each multus non-default network, thus declaring intent of having persistent
+IP allocations for those interfaces.
+This could be further improved by having KubeVirt only create those for multus
+non-default network having the `allowPersistentIPs` knob enabled - KubeVirt
+already reads `NetworkAttachmentDefinition`s. Refactoring it to read those via
+informers would be good, to alleviate the API load.
+
+The IPAM CNI plugin will then react accordingly and if the NAD configuration
+allows it, allocate IPs from its pool - which will be persisted in the
+`IPAMClaim` status. Finally, the CNI will configure the interface with these
+IP addresses.
+
+Please refer to the diagram below to better understand the proposed workflow
+for VM creation:
+1. the user provisions a VM object
+2. the KubeVirt controller creates an IPAMClaim for each multus non-default
+  network in the corresponding VMI spec
+3. OVN-K reacts to the creation of the `IPAMClaim`s. It will generate IP address
+  from the required IP pools, and update the corresponding `IPAMClaim` with the
+  generate IPs.
```

Comment:
I would be clear that:

- OVN-K is not bound to update the IPAMClaim until IPs are actually allocated, which is up to OVN-K when that happens
- Users should not be bound to the status of IPAMClaim for any purpose other than informative
- OVN-K will not check for consistency errors related to the IPAMClaim, like for example, the same IPAMClaim used in two different pods simultaneously.

The general concern is that we are not trying to implement a general purpose IPAM service here, and certainly OVN-K does not intend to be one, so we should be very light on the requirements and have a clear intention that this is not where this is headed.

Reply:
> I would be clear that:
> 
> * OVN-K is not bound to update the IPAMClaim until IPs are actually allocated, which is up to OVN-K when that happens

Agreed.

> * Users should not be bound to the status of IPAMClaim for any purpose other than informative

Makes total sense; that's one of the reasons why having a status sub-resource makes sense here; only OVN-K (or any other CNI allocating the IP) can write to the status - all other components should have (only) read-access to the status.

> * OVN-K will not check for consistency errors related to the IPAMClaim, like for example, the same IPAMClaim used in two different pods simultaneously.

Agreed; if the `IPAMClaim` is not an OVN-Kubernetes resource, it should not be baby-sat by OVN-Kubernetes. 

I don't think it is needed, but, if it turns out to be, we can always derive an error recovery system to address that situation (out of OVN-Kubernetes / common for all CNIs).

> 
> The general concern is that we are not trying to implement a general purpose IPAM service here, and certainly OVN-K does not intend to be one, so we should be very light on the requirements and have a clear intention that this is not where this is headed.

I agree. We pretty much only need a way for the allocation from the light-weight IPAM service to outlive the pod. 

I will make sure this is reflected in the doc.

Reply:
> Makes total sense; that's one of the reasons why having a status sub-resource makes sense here; only OVN-K (or any other CNI allocating the IP) can write to the status - all other components should have (only) read-access to the status.

This has some additional meaning. 

For example, kubevirt should not expect or wait for the IPAMClaim status to be populated before actually adding a pod that uses it. 

Let's say for example that OVNK is not the only component present in the cluster that can honor IPClaims. OVNK would know it has to honor it when it receives a pod request with an IPClaim on one of the networks it handles but not before that.

Reply:
Added the requested information to the doc. Thanks for raising.

> Let's say for example that OVNK is not the only component present in the cluster that can honor IPClaims. OVNK would know it has to honor it when it receives a pod request with an IPClaim on one of the networks it handles but not before that.

I agree with this flow - but I think this is already described in the document. Do you have any suggestion on how I could somehow highlight it ?